mmcls/core/hook/precise_bn_hook.py

```diff
@@ -0,0 +1,189 @@
+    average of per-batch mean/variance instead of the running average.
+
+    Attributes:
+        model (nn.module): The model whose bn stats will be recomputed.
+        loader (DataLoader): PyTorch dataloader._dataloader
+        num_iters (int): number of iterations to compute the stats.
+        logger (:obj:`logging.Logger` | None): Logger for logging.
+            Default: None.
+    """
+    if is_parallel_module(model):
+        parallel_module = model
+        model = model.module
+    else:
+        parallel_module = model
+
+    # get dist info
+    rank, NUM_GPUS = get_dist_info()
+    # Compute the number of minibatches to use, if the size of dataloader is
+    #  less than num_iters, use all the samples in dataloader.
+    num_iter = num_samples // (loader.batch_size * NUM_GPUS)
+    num_iter = min(num_iter, len(loader))
+    # Retrieve the BN layers
+    bn_layers = [
+        m for m in model.modules()
+        if m.training and isinstance(m, (_BatchNorm))
+    ]
+
+    if len(bn_layers) == 0:
+        print_log('No BN found in model', logger=logger, level=logging.WARNING)
+        return
+    print_log(
+        f'{len(bn_layers)} BN found, run {num_iter} iters...', logger=logger)
+
+    # Finds all the other norm layers with training=True.
+    for m in model.modules():
+        if m.training and isinstance(m, (_InstanceNorm, GroupNorm)):
+            print_log(
+                'IN/GN stats will not be updated in PreciseHook.',
+                logger=logger,
+                level=logging.WARNING)
+
+    # Initialize BN stats storage for computing
+    # mean(mean(batch)) and mean(var(batch))
+    running_means = [torch.zeros_like(bn.running_mean) for bn in bn_layers]
+    running_vars = [torch.zeros_like(bn.running_var) for bn in bn_layers]
+    # Remember momentum values
+    momentums = [bn.momentum for bn in bn_layers]
+    # Set momentum to 1.0 to compute BN stats that reflect the current batch
+    for bn in bn_layers:
+        bn.momentum = 1.0
+    # Average the BN stats for each BN layer over the batches
+    if rank == 0:
+        prog_bar = mmcv.ProgressBar(num_iter)
+
+    for data in itertools.islice(loader, num_iter):
+        parallel_module(**data)
+        for i, bn in enumerate(bn_layers):
+            running_means[i] += bn.running_mean / num_iter
+            running_vars[i] += bn.running_var / num_iter
+        if rank == 0:
+            prog_bar.update()
+
+    # Sync BN stats across GPUs (no reduction if 1 GPU used)
+    running_means = scaled_all_reduce(running_means, NUM_GPUS)
+    running_vars = scaled_all_reduce(running_vars, NUM_GPUS)
+    # Set BN stats and restore original momentum values
+    for i, bn in enumerate(bn_layers):
+        bn.running_mean = running_means[i]
+        bn.running_var = running_vars[i]
+        bn.momentum = momentums[i]
+
+
+@HOOKS.register_module()
+class PreciseBNHook(Hook):
+    """Precise BN hook.
+
```

Comment:
May indicate its recommended priority when it is used. For example, its priority should be "xxx" and should be executed before xxx hook after xxx hook

Reply:
done.